src/util/settings_persister.js

```diff
@@ -32,6 +55,14 @@ const debouncedPushConfigToSyncBackend = _.debounce(
             alert(`There was an error trying to push settings to your sync backend: ${error}`)
           );
         break;
+      case 'GitLab':
+        // INFO: Not calling syncBackendClient.createFile is a
```

Comment:
I think this comment might be misleading, because if you look at the `updateConfigForGitlab` function the main reason that it exists is actually to determine whether to call `updateFile()` or `createFile()`. See this comment (which maybe should be promoted to a jsdoc string on the function instead of an inline comment inside it): 
https://github.com/200ok-ch/organice/blob/016d07132658ce8472889f4f3b646b41bc7ab1c8/src/util/settings_persister.js#L26-L27

I think it would make more sense to add the https://github.com/200ok-ch/organice/issues/736 link to this comment:
https://github.com/200ok-ch/organice/blob/016d07132658ce8472889f4f3b646b41bc7ab1c8/src/util/settings_persister.js#L33

Reply:
I did read the code. That's why I've annoated some docs. It's much appreciated that you've documented your rationale so well! :pray: 

I appreciate that GitLab doesn't support creating a file if it doesn't exist. That's not a reason to not adhere to the strategy pattern for sync-backends, though. The reason there's a strategy pattern is exactly that every sync back-end differs in the strategy. However, they all implement the same functionality and in fact even functions. `createFile` for GitLab could do the same sanity check than  `updateConfigForGitlab` does, could it not?

Reply:
> I appreciate that GitLab doesn't support creating a file if it doesn't exist.

Correct me if I'm wrong, but I assume you meant "doesn't support **updating** a file if it doesn't exist".

> The reason there's a strategy pattern is exactly that every sync back-end differs in the strategy.

Definitely agreed, and to your point I agree it's unfortunate that there are so many `switch` statements around backend client calls that make this a leaky abstraction -- another example is the Google Drive backend having three arguments for `createFile()` instead of two.

> createFile for GitLab could do the same sanity check than updateConfigForGitlab does, could it not?

Yes, although that makes me confused about the intended strategy interface. It seems like since there are separate `createFile` and `updateFile` functions that the GitLab implementation is "correct", because otherwise the strategy interface would have a single `createOrUpdateFile` function instead of distinct create vs update functions. I realize the other backends allow updating an existing file via `createFile()` but that seems like misuse of a leaky abstraction as opposed to intended behavior. I'm only basing this off my own inference while looking at the code base since there isn't a spec/contract for the strategy, so I might be way off base.

In theory, it seems like the correct implementation of `debouncedPushConfigToSyncBackend` should be something like one of these two options (rough and untested, just for illustration):
```javascript
// Pre-emptively figure out whether create or update is the right call to make. This makes sense
// if there's an easy way to track this via local state, but I'm not sure how straightforward
// that would be. If the only way to determine if fileExists is by something like
// syncBackendClient.getFileContents(..) then I think the other example is better.
const debouncedPushConfigToSyncBackend = _.debounce(
  (syncBackendClient, contents) => {
    const fileExists = figureThisOut();
    let promise;
    if (fileExists) {
      promise = syncBackendClient.updateFile('/.organice-config.json', contents);
    } else {
      promise = syncBackendClient.createFile('/.organice-config.json', contents);
    }
    promise.catch((error) =>
      alert(`There was an error trying to push settings to your sync backend: ${error}`)
    );
  },
  1000,
  { maxWait: 3000 }
);

// Try updating first because the file is going to exist most of the time this is called, then
// fall back to create if it fails.
// This is likely a better idea than the other example.
const debouncedPushConfigToSyncBackend = _.debounce(
  (syncBackendClient, contents) => {
    syncBackendClient.updateFile('/.organice-config.json', contents).catch(() => {
      syncBackendClient
        .createFile('/.organice-config.json', contents)
        .catch((error) =>
          alert(`There was an error trying to push settings to your sync backend: ${error}`)
        );
    });
  },
  1000,
  { maxWait: 3000 }
);
```

Unfortunately the Google Drive backend breaks the strategy interface completely with the way it takes an extra 'root' argument, so that would have to be changed too, which would require a non-trivial amount of work I imagine. So to be clear I'm not suggesting we make a change like this right now, although in the future it could be a beneficial (albeit likely low priority) improvement.

---

Anyways, all of that was a bit of a tangent to the original point I was trying to make. I think we're talking past each other because there's two issues going on in this section of code:
1. The redundant config syncs issue in #736 
2. The leaky abstraction issue that you just pointed out ([here, for clarity](https://github.com/200ok-ch/organice/pull/734#discussion_r740316521)) and I was responding to in the first half of this message.

What I meant in my original post in this thread was that maybe the code comments should be rearranged a little to clarify which bit of code is dealing with which of the two concerns. I'm not sure if I'm conveying my point very well, so I made a commit to illustrate what I originally meant about rearranging the comments: https://github.com/200ok-ch/organice/pull/734/commits/53455e3d1a747e4f463a3366eaf6c23b98fbc308 . I don't mind dropping the commit if you disagree.

Also I think I'm turning a molehill of a comment into a mountain here, which wasn't my intent.  If I'm still not making sense or you disagree then let's just drop the commit that I added and leave it the way you had it.

Reply:
> Correct me if I'm wrong, but I assume you meant "doesn't support updating a file if it doesn't exist".

Typo. Yes.

> another example is the Google Drive backend having three arguments for createFile() instead of two.

True. Everything Google Drive is unfortunate (of course I mean _only_ in organice *cough*). We've been waiting for aeons and I've personally put a lot of effort into trying to get Google to enable production mode for the [Drive App](https://github.com/200ok-ch/organice/issues/127). So far without success. There's other issues with the Drive integration, too. That it is now also in the way of creating a more solid sync back-end abstraction only compounds the already huge issue around Google Drive.  

So the next step is either a complete cleanup and getting the Drive app to work or just to remove the code. Any contributor who runs into this the next time is welcome to remove it.  I'm not aware that any maintainer is using it. If somebody else wants to step up and fix it, they can still do it.

> Yes, although that makes me confused about the intended strategy interface. It seems like since there are separate createFile and updateFile functions that the GitLab implementation is "correct", because otherwise the strategy interface would have a single createOrUpdateFile function instead of distinct create vs update functions.

I didn't say that the interface is set in stone(; I just wanted to say that it's important to try and keep things as tidy as possible and not deviate if there's an existing pattern unless absolutely needed.

> Also I think I'm turning a molehill of a comment into a mountain here, which wasn't my intent. If I'm still not making sense or you disagree then let's just drop the commit that I added and leave it the way you had it.

Thank you for taking the extra effort of putting some emotional context to your statements above :pray: 

Let's make this simple and just merge. I'm quite sure we're pretty much on the same page on everything(;

Thank you also for putting in the extra effort to further debug the strategy interface and come up with some (pretty solid) mock code. I'll put it into an appropriate issue for safe keeping and another time^^